### PR TITLE
Parse and load binaries

### PR DESCRIPTION
**Pull request by dancrossnyc**
Parse and load the system binaries into the
respective address spaces.

Signed-off-by: Dan Cross <cross@gajendra.net>

### CODE REVIEW

File: theon/src/main.rs [hunk 1]
```diff
@@ -13,21 +13,175 @@
 #![cfg_attr(not(test), no_main)]
 #![cfg_attr(not(test), no_std)]
 
+//!
+//! Theon has two primary responsibilities:
```
Comment (darkstar62) [inline theon/src/main.rs:17]: Thanks for adding this!  I was gonna take a crack at adding some comments myself as I was trying to figure out how all this works.

File: theon/src/main.rs [hunk 2]
```diff
@@ -13,21 +13,175 @@
...
+#[derive(Clone, Copy, Debug)]
+enum BinaryType {
+    Segment,
+    Task,
+}
+
+type BinaryMeta = (&'static str, HPA, BinaryType);
+
+const fn load_addr(offset: usize) -> HPA {
+    let addr = (64 + offset * 32) * MIB;
```
Comment (darkstar62) [inline theon/src/main.rs:51]: Can you add a comment as to how this number is derived?

File: theon/src/main.rs [hunk 3]
```diff
@@ -13,21 +13,175 @@
...
 
+use core::ops::Range;
+
+use arch::{VPageAddr, HPA, MIB, PF4K, V4KA};
+
+type Result<T> = core::result::Result<T, &'static str>;
+
+#[derive(Clone, Copy, Debug)]
+enum BinaryType {
+    Segment,
```
Comment (darkstar62) [inline theon/src/main.rs:44]: Doc comments for these?  What's a Segment vs. Task?

File: theon/src/main.rs [hunk 4]
```diff
@@ -13,21 +13,175 @@
...
+        assert_eq!(vm.start % 4096, 0);
+        assert!(vm.start < vm.end);
+        regions.push(V4KA::new(vm.start)..V4KA::new_round_up(vm.end));
+        headers.push(header);
+    }
+    let base = region.start.vaddress() as *mut u8;
+    let len = region.end.vaddress() - region.start.vaddress();
+    let heap = unsafe { core::slice::from_raw_parts_mut(base, len) };
+    let bump = allocator::BumpAlloc::new(heap);
+    let allocator = || {
```
Comment (darkstar62) [inline theon/src/main.rs:141]: Rename to "allocate" so later calls look like "let mem = allocate().expect();"

File: xtask/src/main.rs [hunk 5]
```diff
@@ -214,6 +214,8 @@ fn qemu(profile: Build) -> Result<()> {
         .arg("-nographic")
         .arg("-cpu")
         .arg("kvm64,+rdtscp,+pdpe1gb,+fsgsbase,+x2apic")
+        .arg("-m")
+        .arg("2048")
```
Comment (darkstar62) [inline xtask/src/main.rs:218]: Is this safe given we're mapping the first 4 GiB of memory for use?

File: x86_64/src/vm.rs [hunk 6]
```diff
@@ -5,13 +5,14 @@
 // license that can be found in the LICENSE file or at
 // https://opensource.org/licenses/MIT.
 
-/// # x86_64 recursive page map implementation
-///
-/// Hypatia uses recursive page tables with side-loading for
-/// address space inspection and manipulation.
+//! # x86_64 recursive page map implementation
```
Comment (darkstar62) [inline x86_64/src/vm.rs:8]: Can you bust up these changes into another commit?

File: theon/src/x86_64/multiboot1.rs [hunk 7]
```diff
@@ -5,6 +5,7 @@
 // license that can be found in the LICENSE file or at
 // https://opensource.org/licenses/MIT.
 
+use crate::end_addr;
```
Comment (darkstar62) [inline theon/src/x86_64/multiboot1.rs:8]: Can you bust this out into another commit?

File: theon/src/main.rs [hunk 8]
```diff
@@ -13,21 +13,175 @@
...
+/// this region, so we can address them via pointers.
 #[cfg_attr(not(test), start, no_mangle)]
 pub extern "C" fn main(mbinfo_phys: u64) -> ! {
-    x86_64::init::start(mbinfo_phys);
+    let multiboot = x86_64::init::start(mbinfo_phys);
+    let crate::x86_64::multiboot1::InitInfo { memory_regions, regions, modules } = multiboot.info();
+    core::mem::drop(memory_regions);
+    uart::panic_println!("end = {:016x}", end_addr());
+    uart::panic_println!("regions: {:#x?}", regions);
+    let bins = modules.iter().find(|m| m.name == Some("bin.a")).expect("found 'bin.a' in modules");
```
Comment (darkstar62) [inline theon/src/main.rs:93]: This archive loading stuff feels like it should be in its own function.

File: theon/src/main.rs (hunk 1 above)
Comment (dancrossnyc) [reply theon/src/main.rs:17]: No problem! I need to do another sweep and add more extensive comments.

File: theon/src/x86_64/multiboot1.rs (hunk 7 above)
Comment (dancrossnyc) [reply theon/src/x86_64/multiboot1.rs:8]: Done.

File: x86_64/src/vm.rs (hunk 6 above)
Comment (dancrossnyc) [reply x86_64/src/vm.rs:8]: Done.

File: theon/src/main.rs (hunk 2 above)
Comment (dancrossnyc) [reply theon/src/main.rs:51]: Done.

File: theon/src/main.rs (hunk 4 above)
Comment (dancrossnyc) [reply theon/src/main.rs:141]: Done.

(Much nicer.)

File: xtask/src/main.rs (hunk 5 above)
Comment (dancrossnyc) [reply xtask/src/main.rs:218]: Yes. Prior to this, it allocated much less (like 256 megs or something) and I was actually writing off the end of RAM loading the binaries. Indeed, I need to add some more error checking here.

File: theon/src/x86_64/multiboot1.rs (hunk 7 above)
Comment (dancrossnyc) [reply theon/src/x86_64/multiboot1.rs:8]: Done.

File: xtask/src/main.rs (hunk 5 above)
Comment (darkstar62) [reply xtask/src/main.rs:218]: I mean, though, won't we still be at risk of writing off the end of memory if we're mapping up to 4GB but only exposing a machine with 2GB?  (at least, without checking)

File: theon/src/main.rs (hunk 8 above)
Comment (dancrossnyc) [reply theon/src/main.rs:93]: I gave it a go, but it actually got pretty messy pretty fast. Would you object to keeping this as-is for now, and then refactoring in a later cleanup?

File: xtask/src/main.rs (hunk 5 above)
Comment (dancrossnyc) [reply xtask/src/main.rs:218]: Oh, and a quick note on the 4GiB thing: there are things in the physical address space that aren't necessarily RAM (like the PCI hole). Mapping those 4GiB lets us move out of 32-bit protected mode with everything still conveniently accessible in the same way it was before turning on paging, and it happens to be reasonably convenient in theon as well.

File: theon/src/main.rs (hunk 8 above)
Comment (darkstar62) [reply theon/src/main.rs:143]: OK.  Maybe add a TODO so we don't forget.

**Review by darkstar62 — APPROVED** [review]

File: theon/src/main.rs (hunk 8 above)
Comment (dancrossnyc) [reply theon/src/main.rs:147]: Done.

File: theon/src/main.rs (hunk 3 above)
Comment (dancrossnyc) [reply theon/src/main.rs:93]: Added comments. It's an internal definition, local just to this module (really just for metadata) so not a doc comment.

File: xtask/src/main.rs (hunk 5 above)
Comment (dancrossnyc) [reply xtask/src/main.rs:218]: Added some more checking here.

(I also didn't realize we were colliding on comments! Too funny.)

File: theon/src/main.rs [hunk 9]
```diff
@@ -13,21 +13,244 @@
...
+
+/// Metadata used in the binary table: the name of the binary,
+/// it's physical load address, and its type (either a segment
+/// or a task).
+type BinaryMeta = (&'static str, HPA, BinaryType);
+
+/// Binaries are loaded in 16MiB regions of physical memory
+/// that are aligned on 32MiB boundaries, starting at 64MiB.
+///
+/// This layout
```
Comment (darkstar62) [inline theon/src/main.rs:105]: Incomplete sentence?

Comment (dancrossnyc) [reply theon/src/main.rs]: Done.

(Removed; subsumed by the description in the file-level comment)

**Review by darkstar62 — APPROVED** [review]

**Pull request merged**